From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Adam Chen <encodeous7@gmail.com>
Date: Mon, 13 Mar 2023 21:23:16 -0400
Subject: [PATCH] Undo add server branding


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 98262d1a6f881ca59f492c2f10d199b091362a96..7886e6d409903f34a13d8c3a21623ab4870ed4dc 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1679,7 +1679,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "PaperTime, by encodeous"; // PaperTime -> Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        return "Paper"; // PaperTime -> Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
